### ISSUE TITLE
Workload ID: SystemD Workload Attestation

### ISSUE THREAD

**Issue by strideynet** [gravitational/teleport#48731]
Today, we have simple Unix Workload Attestation that identifies the PID/UID/GID of the attesting workload. We could enrich this with information from SystemD to determine the service etc. This will provide a more natural way of assigning a SPIFFE ID to a workload.

It'd be nice to enrich the simple Unix Workload Attestation a little further too, e.g, we could capture which binary the process is running, and a hash of that binary.

**Comment by boxofrad** [comment]
📝 **Notes from my initial research**

We can use the [D-Bus API](https://www.freedesktop.org/wiki/Software/systemd/dbus/) (via [godbus](https://github.com/godbus/dbus)) to find a process' unit:

```bash
busctl call org.freedesktop.systemd1 \
  /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager \
  GetUnitByPID u <PID>
```

This will return an object path (e.g. `/org/freedesktop/systemd1/unit/foo_2eservice`) which can used to find other attributes via D-Bus, or decoded into the unit name. Still looking for the docs on the escape sequences, but I know `_2e` represents `.`, for example.

**Comment by boxofrad** [comment]
> we could capture which binary the process is running, and a hash of that binary

@strideynet were you thinking of implementing this by reading `/proc/<pid>/exe`?

Generally, this should be fine because even if you overwrite the binary of a running process (e.g. during a rolling upgrade), `exe` is a symlink to the original binary's **inode** rather than just its path, but it won't necessarily work if the binary is on a network filesystem where there are no guarantees that inodes are stable.

I discovered this by accident when trying to replace a binary in the `/vagrant` mount of a Vagrant VMWare machine 😅

So we should probably call this out in the documentation at least. Similarly, hashing the binary is probably more useful for statically-linked applications.

**Comment by strideynet** [comment]
> @strideynet were you thinking of implementing this by reading /proc/<pid>/exe?

Yeah I think so. I didn't go in too deep in my research on this. Those limitations sound fine as long as we document them. I imagine this would be more about auditing than authorization. For those who want better controls, I think they probably ought to explore container signing etc.
